src/lib/updating-element.ts

```diff
@@ -0,0 +1,503 @@
+/**
+ * @license
+ * Copyright (c) 2017 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+/**
+ * Converts property values to and from attribute values.
+ */
+type AttributeProcessor = {
+  /**
+   * Deserializing function called to convert an attribute value to a property value.
+   */
+  fromAttribute?(value: string): unknown,
+  /**
+   * Serializing function called to convert a property value to an attribute value.
+   */
+  toAttribute?(value: unknown): string|null
+};
+
+/**
+ * Defines options for a property accessor.
+ */
+export interface PropertyOptions {
+  /**
+   * Describes how and if the property should becomes an observedAttribute.
+   * If explicitly false, the property will not be observed. Otherwise if true
+   * or absent, the lowercased property name is observed, and if a string
+   * observe that value.
+   */
+  attribute?: boolean|string;
+  /**
+   * Describes how to convert a property value to and from an attribute. If
+   * the value is a function, it's calle to deserialize an attribute value into
+   * a property value. If it's an AttributeProcessor, the `fromAttribute` value
+   * is used to deserialize. If the `reflect` option is also true, then
+   * `toAttribute` value is used to serialize the property value to an attribute.
+   */
+  type?: AttributeProcessor|Function;
+  /**
+   * Describes if setting the property should be reflect to the attribute value.
+   * If true, then if present the `type.toAttribute` function is used to serialize
+   * the value; otherwise, the property value is used.
+   */
+  reflect?: boolean;
+  /**
+   * Describes if setting a property should trigger invalidation and updating.
+   * This function takes the new and oldValue and returns true if invalidation
+   * should occur. If not present, a strict identity check is used.
+   */
+  shouldInvalidate?(value: unknown, oldValue: unknown): boolean;
```

Comment:
a motivating example on when you don't want something to invalidate would be quite useful

Reply:
Done